### DISCUSSION TITLE
[FEATURE REQUEST] Orchid RAMQuest 8/16

### DISCUSSION THREAD

**Discussion by ppgrainbow** [86Box/86Box#4528] [Ideas]
I would like to see support for the Orchid RAQMQuest 8/16 memory expansion card emulation for PC/XT, AT, 386 and 486-based PCs and up! It's the first and only documented 16-bit ISA expansion card that can expand up to 32 MB on the expansion board. Here is the manual on how the RAMQuest 8/16 expansion card and how to use it:

https://www.minuszerodegrees.net/manuals/Orchid/Orchid%20-%20RamQuest%208_16%20-%20Users%20Manual.pdf

Here is some documentation regarding the RAMQuest 8/16 memory expansion card:
https://theretroweb.com/expansioncard/documentation/50218.pdf

Lastly, here are the drivers for the RAMQuest 8/16 expansion card dated June 14, 1990:
https://www.minuszerodegrees.net/manuals/Orchid/Orchid%20-%20RamQuest%208_16%20-%20Drivers.zip

While this memory expansion board can provide EMS to 80286-based machines and up, EMS RAM and UMBs can be provided on XT-based machines; however, XMS memory will not be able on XT-based machines, because only 1 MB of RAM will be addressable to the main board.

The only drawback is that 386 and 486-based motherboards use **70 to 90 nanoseconds** speed timing for main board RAM and since RAMQuest 8/16 expansion memory card requires the use of 100 nanosecond speed RAM, you can expect the timing of the RAM speed to be **10 to 30 nanoseconds** slower - and 86Box doesn't emulate speed timings of the RAM.

If this were to be implemented in 86Box, the start address and memory size will need to be configured in 512 KB increments up to 32 MB.

For 386 (DX) and 486-based emulated PCs that have 32 MB of RAM, the start address would have to be set at 32768 and with the maximum size is set to 32768, up to 64 MB of RAM (32 MB of system RAM + 32 MB of RAM on the expansion card) will be detected.

If you have any questions or comments about my feature request, let me know. :)

**Comment by Torinde** [comment]
Can you use multiple of those 32MB cards together - for 64MB (or more) EMS?

Do you have an idea how [F8680A](https://github.com/joncampbell123/dosbox-x/discussions/5261) (C&T186) is addressing its 64MB/26-bits?

1. custom addressing mode (like NEC V33)
2. non-standard EMS implementation
3. something else

**Comment by Fenix770** [comment]
Interesting. 
I guess more EMS memory can be used patching the EMS driver. But anything more than 4mb would be a waste of RAM, and slow to the machine to detect it. 

The EMS memory provided by the cards cannot be "seen" by the CPU in the same way as extended memory. This memory is controlled by the EMS hardware, and the CPU only "sees" the EMS page frame.

Except when a 386 with a 386 memory manager is used, that emulates the EMS using the V86 mode of the CPU.
Or if a memory manager unit is installed on a 286, like the All-Chargecard or the Sota Pop Card. 

These were adapters that worked directly with the 286 CPU, and it did some functions that only 386 machines could do. 
Like do some multitasking on Deskview. This adapter allowed the CPU to open several DOS apps, in a way similar to Windows 2.1x/386 on 386 machines.

These adapters were sold as upgrades for 286 users until 1991. But in late 1991 the prices of 386 SX CPUs dropped and in 1992 Cyrix launched the 486 SLC/SxR CPUs. 
So in theory you can upgrade a 1989-90 286 machine with a 486 SxR/2 33 MHz (if the board supports a 16 MHz bus ) or a 486 SLC/20 or a SxR/2 40 MHz (if supports a 20Mhz bus).

I found some info some years ago. 
Some info here: https://web.archive.org/web/20230504111227/https://sites.google.com/site/misterzeropage/

This guy managed to install this card on a NEC V20 machine, and did some RAM backfilling, that can be used as large page frame on Windows 2.x and  3.0 in real mode.

It is also possible to map some UMBs using QRAM. 

But I found also some generic information included on Qualitas Move'em (808x/286 version of 386Max) 
It mentions the 16/32 version of the Ramquest not the 8/16 but can map very little UMBs. 32kb to 64kb depending on the video card installed (CGA,EGA or VGA).

[EMSGUIDE.TXT](https://github.com/user-attachments/files/18633891/EMSGUIDE.TXT)

**Reply by Torinde** [reply]
What do you think [F8680](https://github.com/86Box/86Box/discussions/4922) "maximum local system memory that is directly managed is 4MB" refers to? I understand where a 1MB or 32MB limit comes from, but not 4MB?

As for the 64MB total being above the LIM4.0 maximum of 32MB - is it possible that F8680A allow simultaneous usage of:
- up to 32MB [Mappable conventional memory](https://github.com/86Box/86Box/discussions/4054#discussioncomment-12044551) for Windows applications
- up to 32MB Expanded memory for DOS applications

**Comment by Fenix770** [comment]
As for 4Mb you may be speaking of a 80186/80188. These CPUs supported protected mode. 
The conventional memory is always 640kb not more since it works in the real mode compatible with the 8086/88.

1. "Mappable" means that there is a "window" be on the Sort 64kb page frame (like D000-DFFF) or large page frame (an area shared with the conventional memory area like 4000-9FFF) can be used to bank-switch to the EMS memory pool located *outside* the conventional (up to 640kb) or upper memory area (located somewhere in the A000-FFFF).

That strategy was developed to be compatible with the real mode operation of the DOS apps at the time and the 808x/286 architecture.
Other way was to use a protected mode OS, on 286+, like OS/2 and use the extended memory directly.

2. The 32mb limit is an arbitrary  limit established by the LIM 4.0 standard launched in 1987. Many high-end cards LIM EMS cards had a hardware limit of 8mb and even high-end systems used to have 4mb of LIM EMS 4.0 RAM. 

With a 386 memory manager, like QEMM, you can have more than 32mb of LIM EMS RAM.  However, that amount of RAM had no use except may be if you run several programming apps at the same time with a software like Deskview.

3. The "mappable" memory was a solution to three issues, 

A. The 286 and lesser computers were the most common PC at the time.

B. Backward compatibility, DOS apps. (A same DOS app should be runnable on 8086+)

C. Bank-switching technique were faster when large chunks of RAM were moved on 286. There was a comparison on 286/10, in 1989 between  Excel for OS/2 vs Excel for Windows 2.11/286. The Windows version was faster when loading large files (2 MB)

Drawback 

D. Both hardware and software must support the large page frame to take advantage of it. 

I discovered that some Windows 3.0 games can run out of memory if Windows is in real mode with large page frame enabled, because these were not programmed to use the large page frame.

The real mode of Windows 3.0 was meant to maintain compatibility with Windows 2.x apps.

4. Then in 1990 some apps started to use the 16-bit protected mode to use extended memory on the 286+, (like Lotus 1-2-3 v3.0). 
And of course Windows 3.0 in standard and 386 enhanced mode

So the extended memory, that had almost no use, became useful. And the hardware EMS was deprecated.


Btw: 

-The LIM 4.0 standard states that the EMS memory can be accessed without a page frame, but still required by almost all DOS apps for comaptibility reasons.

-I tested DR-DOS 6 EMM386.SYS, and it does not support the large page frame, however is possible to run several apps with the TaskMax utility. 

-I think the LIM EMS standard was developed at about the same time as the 386 (in late 1985). Because the high price of the 386 and the need to maintain compatiblity with DOS apps.

- I also think that the V86 mode on 386+ can perform more functions and faster than the hardware EMS.

**Reply by Torinde** [reply]
> As for 4Mb you may be speaking of a 80186/80188. These CPUs supported protected mode.

80186/188 did NOT support protected mode. 80286 introduced protected mode. Also, 80286 extended memory limit is 16MB, not 4MB.

The F8680A is not a regular x86 and has extensive memory controller features (see from [page 20](https://datasheets.chipdb.org/Chips/F8680A_Rev_2_0.pdf) onwards). It supports Expanded memory/EMS (similar to the EMS boards), but somehow also goes beyond 32MB. The 4MB is a separate thing (unclear to me), maybe related to its four "Memory Window Registers"?

The "Mappable conventional memory" and "Expanded memory" are specific terms mentioned in the Intel [Above Board documentation](https://www.minuszerodegrees.net/manuals/Intel/Intel%20-%20Above%20Board%20Plus%208%20and%20Above%20Board%20Plus%208%20IO%20-%20Installation%20Guide%20-%20May%201989.pdf). See page 2-5/2-6 and appendix E:
![image](https://github.com/user-attachments/assets/6fc37eb4-0fb4-4093-a72d-a8cda60070a7)

You write about "large page frame" - is that another label for the "Mappable conventional memory" (e.g. the twice bigger window located inside the lower 640KB)?

> LIM 4.0 standard states that the EMS memory can be accessed without a page frame

Can you please point to where in [LIM4.0](https://web.archive.org/web/20040718235414/http://nersp.nerdc.ufl.edu/~esi4161/files/dosman/chap10) to read about that?

Three questions remain:
- [Intel Above Board](https://github.com/86Box/86Box/discussions/4054#discussioncomment-12044551): what will happen with multiple boards if 4x14MB=56MB are installed?
- F8680A: in what way does it use the 32MB-to-64MB space?
- F8680A: what does it mean "maximum local system memory that is directly managed" (4MB)?

**Comment by Fenix770** [comment]
I think the 80186/88 may be *not* IBM PC compatible. 

1. This Wikipedia claims that it have 20-bit for memory address. So it can "see" only 1mb at a time, and the rest may require some kind of bank switching to access it.   
https://es.wikipedia.org/wiki/Intel_80186_y_80188


The 286 or latter can only use more than 1mb linear memory in protected mode. (For "linear" I mean non bank switched)

I read some time ago that some other non PC machines used the bank switching techniques to work around the CPU mapping limitations, for example early VGA cards used it to access more that 256kb VRAM (like the Trident 8800, 8900 A and B).

2. I think there is no limit to the amount to LIM EMS, except the hardware limit, if I recall correctly the LIM XMS 3.0 (not EMS) has a limit of 64Mb, on 386 systems QEMM can convert it to LIM EMS 4.0, even if the specs says 32Mb is the limit.

I guess if the hardware LIM EMS driver is patched, and you have four cards that supports 32mb on each one, then you may have 128 mb of LIM EMS.

3. I don't know what "maximum local system memory that is directly managed". But I guess is the maximum memory the CPU can "see"  at any given time. 
For example on the LIM EMS 3.2 spec

A. the page frame is 64kb (memory windows and the maximum portion of memory size the 808x CPUs can read or write, hence the 16-bit internal bus)

B. The "window" is 256kb (is the biggest block the card can see, before doing a bank-switch). 
This was expanded to 1Mb to match the 20-bit bus of the real mode on the LIM 4.0. 

So if the card is fully LIM 4.0 capable it can store up to 1Mb of executable program code, not just data like the LIM EMS 3.2 can do)

But many cards were just LIM 3.2 at hardware level with a LIM 4.0 device driver since the spec allows emulating in software the extra register sets (up to 64), using only one.


I did'n find the reference to accessing the EMS memory without the page frame. So it may be a thing that latter 386 memory managers added, outside the spec.

Here is the LIM 4.0 info
https://www.phatcode.net/res/218/files/limems40.txt

**Reply by Torinde** [reply]
80186/188 (and F8680) is fully x86 compatible, thus [computers using](https://en.wikipedia.org/wiki/Gateway_HandBook) it _can_ be IBM PC compatible. The documentation of F8680 explicitly states IBM PC compatibility. And yes, the documentation also confirms it uses bank switching.

F8680 rev1.1 documentation seems to state 32MB EMS?
> Once mapped, segments can be bank-switched for implementation of EMS memory and a PCMCIA memory card interface. ...
A total of 32MB memory can be addressed through the 26-bit address bus ADR25:0 (the 32MB indicated when ADR25=1 is addressable only as PCMCIA memory).

While rev2.0 (for F8680**A**) seems to state 64MB?:
> Once mapped, segments can be bank-switched for implementation of EMS memory and a PCMCIA memory card interface. ...
A total of 64MB memory can be addressed through the 26-bit address bus ADR25:0.

And the product is from the 1990s, way after EMS4.0 (32MB). EMS3.2 is 8MB and only the older and least used one EMS3.0 is 4MB. So, I don't think 4MB limit is related to the EMS3.0 limit. Also, I don't think F8680 supports XMS or HMA (do you see such in the docs?) and those are 286 features.

> if the hardware LIM EMS driver is patched

I agree with you this may be the explanation. Where does the 32MB limit in EMS4.0 come from? What has to be changed to double it? Does that match the details in the documentation ([page24-51](https://datasheets.chipdb.org/Chips/F8680A_Rev_2_0.pdf))?

> EMS memory without the page frame

I found that now: "[In fact, you need no page frame at all.](https://www.betaarchive.com/wiki/index.php/Microsoft_KB_Archive/66420#Expanded_Memory)"

**Comment by Torinde** [comment]
What happens if you have two different EMS cards, with separate drivers - do you get two separate "memory windows" into two separate EMS pools? Or they'll clash and won't work together? Or there is a coordination mechanism to use them jointly?

**Comment by Fenix770** [comment]
I didn't read the  F8680, but I don't think any machine can map 64mb on real mode. I speak about the LIM EMS RAM/ spec. 
The EMS 3.2; 4.0 software specs are just that specs. The real memory limit is stablished by the hadrware.

A. The confusion may be that

1.hardware EMS memory limit is a bank-switched memory managed by the memory card and the CPU only write/read the page frame.
2. Other type of linear memory be conventional (up to 1mb at 0000-FFFF region)  at least on 808x+ or extended memory beyond 1mb.
3. If the 186/188 can map more than 1mb in real mode then it may be not compatible with existing software when operating in that mode.


B. I'm not a programmer but there are some examples already emulated in 86Box, like the Lo-Tech EMS card.
The first card has a 2mb limit, and someone created another with 4mb support and a patched EMS driver.
https://github.com/hkzlab/ISA_EMS_4Mb

C. You can use only one page frame for the EMS bank-switching, in most cases the card need a port for each 2mb, so if more memory is usesd more ports are needed.  

Example: the emulated Everex RAM Deluxe 3000 emulated in 86box. supports 3mb. One port would be 208/209 for the first 2mb and the a 218/219 for the 2-3mb memory segment.

D. You cannot combine incompatible EMS cards. The EMS driver will refuse to load if another one is present. And at hardware level the EMS cards are implemeted differently, so one EMS driver may not work on other EMS cards of other manufacturers.

This is because the LIM EMS spec is the same at software level but each card works differently at hardware level.

**Reply by Torinde** [reply]
Intel 186/188 can't map above 1MB. As for F8680 - yes, that is the other potential explanation - it may have some "custom addressing mode" for that, but from what I read it seems to be more about bank switching than special CPU modes.

Please take a look at the [F8680 docs](https://github.com/86Box/86Box/discussions/4528#discussioncomment-12053331) an share your opinion on:
- F8680(A): what does it mean "maximum local system memory that is directly managed" (4MB)?
- F8680A: in what way does it use the 32MB-to-64MB space?

**Reply by Torinde** [reply]
Assuming F8680A has similar approach to [NEC Vadem VG230](https://github.com/joncampbell123/dosbox-x/issues/5442#issuecomment-2638258118) - probably it uses for EMS "a compatible superset of LIM 4.0".

Do you think that when emulating 286 protected mode (16MB XMS), it can use 48MB EMS in addition?

**Reply by Torinde** [reply]
The answer for the 4MB: refers to the [RAS and CAS signals needed to drive the memory](https://github.com/joncampbell123/dosbox-x/issues/5442#issuecomment-2638317873) directly. More then 4MB needs external circuitry.

**Comment by OBattler** [comment]
What we need is the F8680 PC/CHIP Programmer's Reference Manual which is, however, currently unavailable.

**Reply by Torinde** [reply]
Is the [Vadem VG230](https://web.archive.org/web/20230322201758/http://www.amphus.com/chips/pdf/230dm.pdf) documentation sufficient to emulate its own 64MB LIM4.0 compatible superset? You already have the NEC 286 real mode CPU.
